torch/_inductor/compile_fx.py

```diff
@@ -219,88 +225,649 @@ def remove_unaligned_input_idxs(inputs, static_input_idxs):
+            return weak_ref() is not None
+
+        return [pytree.tree_map(is_live, outputs) for outputs in output_weakrefs]
+
+    @staticmethod
+    def liveness_delta(
+        prev: List[List[bool]], curr: List[List[bool]]
+    ) -> List[Tuple[int, int]]:
+        dead_indices = []
+        assert len(prev) <= len(curr)
+        for i, (outputs1, outputs2) in enumerate(zip(prev, curr)):
+            assert len(outputs1) == len(outputs2)
+            for j, (output1, output2) in enumerate(zip(outputs1, outputs2)):
+                if output1 != output2:
+                    dead_indices.append((i, j))
+
+        return dead_indices
+
+    @staticmethod
+    def check_liveness(indices: List[Tuple[int, int]], output_refs: List[List[bool]]):
+        for i, j in indices:
+            if output_refs[i][j]() is not None:
+                return False
+        return True
+
+    def valid_end_of_execution(self):
+        return len(self.executed_tape) == len(
+            self.recorded_tape
+        ) and self.check_liveness(
+            self.outputs_live_at_last_tape, self.executed_outputs_weakrefs
+        )
+
+    def reset_execution_state(self):
+        self.executed_tape = []
+        self.executed_outputs_weakrefs = []
+
+    def valid_end_of_recording(self):
+        for outputs in self.recorded_outputs_weakrefs:
+            for out in outputs:
+                if out is not None and out() is not None:
+                    return False
+
+        live_outputs = []
+        final_recorded_liveness = self.recorded_liveness_after_graph[-1]
+        for i in range(len(final_recorded_liveness)):
+            for j in range(len(final_recorded_liveness[i])):
+                if final_recorded_liveness[i][j]:
+                    live_outputs.append((i, j))
+
+        self.outputs_live_at_last_tape = live_outputs
+        return True
+
+    def is_cuda_graph_recorded_tensor(self, tensor):
+        cuda_graph_managed_tensors = []
+        for i in range(len(self.recorded_outputs_weakrefs)):
+            for j in range(len(self.recorded_outputs_weakrefs[i])):
+                output = self.recorded_outputs_weakrefs[i][j]
+                if output is not None and output() is tensor:
+                    return True
+
+        return False
+
+
+class TapeManagerContainer(object):
+    """
+    Manages the lifetime of the tape manager. Like `PrivatePool` in cuda caching allocator,
+    the tape and its corresponding memory pool should be kept alive as long as any outstanding
+    graph or tensor which is an output of a graph remains alive.
+    """
+
+    def __init__(self):
+        self.tape_manager = None
+        self.alive_cudagraphify_objs = 0
+
+        # only set in the case where the CudaGraphify objects die while their
+        # tensor outputs remain live
+        self.live_tensors_count = 0
+        # Used to keep alive the memory pool if the CudaGraph object dies
+        self.graph = None
+        self.lock = threading.Lock()
```

Comment:
it suuure would be helpful to have types on these, especially since you seem to be monkeying around with this values from outside of the class directly